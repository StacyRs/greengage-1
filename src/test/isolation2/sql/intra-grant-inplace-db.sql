-- GRANT's lock is the catalog tuple xmax.  GRANT doesn't acquire a heavyweight
-- lock on the object undergoing an ACL change.  In-place updates, namely
-- datfrozenxid, need special code to cope.

CREATE ROLE regress_temp_grantee;

--start_ignore
DROP TABLE IF EXISTS frozen_witness;
--end_ignore

3: CREATE TEMPORARY TABLE frozen_witness (x xid) distributed by (x);
-- observe datfrozenxid
3: INSERT INTO frozen_witness SELECT datfrozenxid FROM pg_database WHERE datname = current_catalog;
1: BEGIN;
-- heap_update(pg_database)
1: GRANT TEMP ON DATABASE isolation2test TO regress_temp_grantee;
-- inplace update
2&: VACUUM (FREEZE);
3: INSERT INTO frozen_witness SELECT datfrozenxid FROM pg_database WHERE datname = current_catalog;
1: COMMIT;
2<:
-- Save the result in an environment variable.
-- We get the raw xid to be sure that the age in the next query will be executed on the coordinator.
3: @post_run 'TOKEN=`echo "${RAW_STR}" | awk \'NR==3\' | awk \'{print $1}\'` && echo ""' : SELECT min(x::varchar::int) FROM frozen_witness;
-- observe datfrozenxid
3: @pre_run 'echo "${RAW_STR}" | sed "s#@TOKEN#${TOKEN}#"': SELECT 'datfrozenxid retreated' FROM pg_database WHERE datname = current_catalog AND age(datfrozenxid) > age('@TOKEN'::xid);

REVOKE ALL ON DATABASE isolation2test FROM regress_temp_grantee;
DROP ROLE regress_temp_grantee;

1q:
2q:
3q:
