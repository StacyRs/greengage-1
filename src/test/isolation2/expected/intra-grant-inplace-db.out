-- GRANT's lock is the catalog tuple xmax.  GRANT doesn't acquire a heavyweight
-- lock on the object undergoing an ACL change.  In-place updates, namely
-- datfrozenxid, need special code to cope.

CREATE ROLE regress_temp_grantee;
CREATE

--start_ignore
DROP TABLE IF EXISTS frozen_witness;
DROP
--end_ignore

3: CREATE TEMPORARY TABLE frozen_witness (x xid) distributed by (x);
CREATE
-- observe datfrozenxid
3: INSERT INTO frozen_witness SELECT datfrozenxid FROM pg_database WHERE datname = current_catalog;
INSERT 1
1: BEGIN;
BEGIN
-- heap_update(pg_database)
1: GRANT TEMP ON DATABASE isolation2test TO regress_temp_grantee;
GRANT
-- inplace update
2&: VACUUM (FREEZE);  <waiting ...>
FAILED:  Forked command is not blocking; got output: VACUUM
3: INSERT INTO frozen_witness SELECT datfrozenxid FROM pg_database WHERE datname = current_catalog;
INSERT 1
1: COMMIT;
COMMIT
2<:  <... completed>
FAILED:  Execution failed
-- Save the result in an environment variable.
-- We get the raw xid to be sure that the age in the next query will be executed on the coordinator.
3: @post_run 'TOKEN=`echo "${RAW_STR}" | awk \'NR==3\' | awk \'{print $1}\'` && echo ""' : SELECT min(x::varchar::int) FROM frozen_witness;

-- observe datfrozenxid
3: @pre_run 'echo "${RAW_STR}" | sed "s#@TOKEN#${TOKEN}#"': SELECT 'datfrozenxid retreated' FROM pg_database WHERE datname = current_catalog AND age(datfrozenxid) > age('@TOKEN'::xid);
 ?column?               
------------------------
 datfrozenxid retreated 
(1 row)

REVOKE ALL ON DATABASE isolation2test FROM regress_temp_grantee;
REVOKE
DROP ROLE regress_temp_grantee;
DROP

1q: ... <quitting>
2q: ... <quitting>
3q: ... <quitting>
